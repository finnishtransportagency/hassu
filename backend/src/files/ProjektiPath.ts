import {
  HyvaksymisPaatosVaihe,
  HyvaksymisPaatosVaiheJulkaisu,
  NahtavillaoloVaihe,
  NahtavillaoloVaiheJulkaisu,
  Vuorovaikutus,
} from "../database/model";

export abstract class PathTuple {
  protected readonly parent: PathTuple;

  constructor(parent?: PathTuple) {
    // TODO Varmista että tämä on ok
    // eslint-disable-next-line @typescript-eslint/ban-ts-comment
    // @ts-ignore
    this.parent = parent;
  }

  abstract get yllapitoPath(): string;

  abstract get publicPath(): string;

  get publicFullPath(): string {
    return this.parent ? this.parent.publicPath + "/" + this.publicPath : "";
  }
}

export class ProjektiPaths extends PathTuple {
  static PATH_NAHTAVILLAOLO = "nahtavillaolo";
  static PATH_HYVAKSYMISPAATOS = "hyvaksymispaatos";
  private readonly oid: string;

  public constructor(oid: string) {
    super();
    this.oid = oid;
  }

  get yllapitoPath(): string {
    return `yllapito/tiedostot/projekti/${this.oid}`;
  }

  get publicPath(): string {
    return `tiedostot/suunnitelma/${this.oid}`;
  }

  get publicFullPath(): string {
    return this.publicPath;
  }

  vuorovaikutus(vuorovaikutus: Vuorovaikutus): VuorovaikutusPaths {
    return new VuorovaikutusPaths(this, vuorovaikutus);
  }

  nahtavillaoloVaihe(nahtavillaoloVaihe: NahtavillaoloVaihe | NahtavillaoloVaiheJulkaisu): PathTuple {
    return new NahtavillaoloVaihePaths(this, nahtavillaoloVaihe);
  }

  hyvaksymisPaatosVaihe(hyvaksymisPaatosVaihe: HyvaksymisPaatosVaihe | HyvaksymisPaatosVaiheJulkaisu): HyvaksymisPaatosVaihePaths {
    return new HyvaksymisPaatosVaihePaths(this, ProjektiPaths.PATH_HYVAKSYMISPAATOS, hyvaksymisPaatosVaihe);
  }
}

class VuorovaikutusPaths extends PathTuple {
  private vuorovaikutus: Vuorovaikutus;

  constructor(parent: PathTuple, vuorovaikutus: Vuorovaikutus) {
    super(parent);
    this.vuorovaikutus = vuorovaikutus;
  }

  get yllapitoPath(): string {
    return "suunnitteluvaihe/vuorovaikutus_" + this.vuorovaikutus.vuorovaikutusNumero;
  }

  get publicPath(): string {
    return "suunnitteluvaihe/vuorovaikutus_" + this.vuorovaikutus.vuorovaikutusNumero;
  }

  get aineisto(): VuorovaikutusAineisto {
    return new VuorovaikutusAineisto(this);
  }
}

class VuorovaikutusAineisto extends PathTuple {
  constructor(parent: PathTuple) {
    super(parent);
  }

  get publicPath(): string {
    return this.parent.publicPath + "/" + "aineisto";
  }

  get yllapitoPath(): string {
    return this.parent.yllapitoPath + "/" + "aineisto";
  }

  get publicFullPath(): string {
    return this.parent.publicFullPath + "/" + "aineisto";
  }
}

class NahtavillaoloVaihePaths extends PathTuple {
  private nahtavillaoloVaiheId: number;

  constructor(parent: PathTuple, nahtavillaoloVaihe: NahtavillaoloVaihe | NahtavillaoloVaiheJulkaisu) {
    super(parent);
    this.nahtavillaoloVaiheId = nahtavillaoloVaihe.id;
  }

  get yllapitoPath(): string {
    return ProjektiPaths.PATH_NAHTAVILLAOLO + "/" + this.nahtavillaoloVaiheId;
  }

  get publicPath(): string {
    return ProjektiPaths.PATH_NAHTAVILLAOLO;
  }
}

class VersionedPaths<T extends { id: number }> extends PathTuple {
  private id: number;
  private folder: string;

  constructor(parent: PathTuple, folder: string, versionedElement: T) {
    super(parent);
    this.folder = folder;
    this.id = versionedElement.id;
  }

  get yllapitoPath(): string {
    return this.folder + "/" + this.id;
  }

  get publicPath(): string {
    return this.folder;
  }
}

class HyvaksymisPaatosVaihePaths extends VersionedPaths<HyvaksymisPaatosVaihe | HyvaksymisPaatosVaiheJulkaisu> {
  get paatos(): PathTuple {
    return new (class extends PathTuple {
      get publicPath(): string {
        return this.parent.publicPath + "/" + "paatos";
      }

      get yllapitoPath(): string {
        return this.parent.yllapitoPath + "/" + "paatos";
      }
    })(this);
  }
}
