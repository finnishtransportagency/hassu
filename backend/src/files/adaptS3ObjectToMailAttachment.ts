import { GetObjectCommandOutput } from "@aws-sdk/client-s3";
import Mail from "nodemailer/lib/mailer";
import { Readable } from "stream";
import { log } from "../logger";

export default function adaptS3ObjectOutputToMailAttachment(
  { ContentType, Body }: GetObjectCommandOutput,
  key: string
): Mail.Attachment | undefined {
  if (!(Body instanceof Readable)) {
    log.error("Liitetiedoston sisällössä ongelmia");
    return undefined;
  }
  const contentType = !ContentType || ContentType == "null" ? "application/octet-stream" : ContentType;
  const getFilenamePartFromKey = (path: string) => path.substring(path.lastIndexOf("/") + 1);
  return {
    filename: getFilenamePartFromKey(key),
    contentDisposition: "attachment",
    contentType: contentType ?? "application/octet-stream",
    content: Body,
  };
}
