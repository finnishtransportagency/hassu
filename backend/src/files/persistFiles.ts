import { fileService } from "./fileService";
import { assertIsDefined } from "../util/assertions";
import { localDateTimeString } from "../util/dateUtil";
import { LadattuTiedosto } from "../database/model";
import { LadattuTiedostoTila, AsiakirjaTyyppi } from "hassu-common/graphql/apiModel";
import { log } from "../logger";

type KasiteltyTiedosto = LadattuTiedosto & { fileWasRemoved: boolean; fileWasPersisted: boolean };
export async function persistLadattuTiedosto({
  oid,
  ladattuTiedosto,
  targetFilePathInProjekti,
  poistetaan,
  asiakirjaTyyppi,
}: {
  oid: string;
  ladattuTiedosto: LadattuTiedosto;
  targetFilePathInProjekti: string;
  poistetaan: boolean;
  asiakirjaTyyppi?: AsiakirjaTyyppi;
}): Promise<KasiteltyTiedosto> {
  let fileWasRemoved = false;
  let fileWasPersisted = false;
  if (!ladattuTiedosto.tuotu || ladattuTiedosto.tila === LadattuTiedostoTila.ODOTTAA_PERSISTOINTIA) {
    log.info("persistoidaan tiedosto", ladattuTiedosto);
    const uploadedFile: string = await fileService.persistFileToProjekti({
      uploadedFileSource: ladattuTiedosto.tiedosto,
      oid,
      targetFilePathInProjekti,
      asiakirjaTyyppi,
    });

    const fileName = uploadedFile.split("/").pop();
    assertIsDefined(fileName, "tiedostonimi pitäisi löytyä aina");
    ladattuTiedosto.tiedosto = uploadedFile;
    ladattuTiedosto.nimi = fileName;
    ladattuTiedosto.tuotu = localDateTimeString();
    ladattuTiedosto.tila = LadattuTiedostoTila.VALMIS;
    fileWasPersisted = true;
  } else if (poistetaan) {
    log.info("poistetaan tiedosto", ladattuTiedosto);
    // Deletoi tiedosto
    await fileService.deleteYllapitoFileFromProjekti({
      oid,
      filePathInProjekti: ladattuTiedosto.tiedosto,
      reason: "Käyttäjä poisti tiedoston",
    });
    fileWasRemoved = true;
  } else {
    log.info("ignoroidaan tiedosto", ladattuTiedosto);
  }

  return { ...ladattuTiedosto, fileWasRemoved, fileWasPersisted };
}
