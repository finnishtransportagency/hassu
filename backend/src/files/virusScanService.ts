import { invokeLambda } from "../aws/lambda";
import { parameters } from "../aws/parameters";
import { log } from "../logger";
import { ProjektiPaths } from "./ProjektiPath";
import { config } from "../config";
import { fileService } from "./fileService";
import { LiitteenSkannausTulos } from "hassu-common/graphql/apiModel";
import { getS3Client } from "../aws/client";
import { GetObjectTaggingCommand } from "@aws-sdk/client-s3";

class VirusScanService {
  async runScanOnFile(bucketName: string, key: string) {
    try {
      const virusScannerLambdaArn = await parameters.getParameter("VirusScannerLambdaArn");
      if (virusScannerLambdaArn) {
        const payload = {
          Records: [
            {
              s3: {
                bucket: {
                  name: bucketName,
                },
                object: {
                  key,
                },
              },
            },
          ],
        };
        log.info("Käynnistetään virusskannaus", { payload });
        await invokeLambda(virusScannerLambdaArn, false, JSON.stringify(payload));
      }
    } catch (e) {
      log.error(e);
      // Älä palauta virhettä kansalaiselle käyttöliittymään
    }
  }

  async getVirusScanResult(path: ProjektiPaths, liitteet: string[] | null | undefined): Promise<LiitteenSkannausTulos | undefined> {
    if (liitteet && liitteet.length > 0) {
      const s3 = getS3Client();
      let result = true;
      for (const liite of liitteet) {
        const tagResponse = await s3.send(
          new GetObjectTaggingCommand({
            Bucket: config.yllapitoBucketName,
            Key: fileService.getYllapitoPathForProjektiFile(path, liite),
          })
        );
        const virusscan = tagResponse.TagSet?.filter((tag) => tag.Key == "viruscan").pop();
        if (virusscan) {
          result = result && virusscan.Value === "clean";
        } else {
          return undefined;
        }
        if (!result) {
          break;
        }
      }
      return result ? LiitteenSkannausTulos.OK : LiitteenSkannausTulos.SAASTUNUT;
    }
  }
}

export const virusScanService = new VirusScanService();
