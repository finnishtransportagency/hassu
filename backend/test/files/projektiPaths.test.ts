/* tslint:disable:only-arrow-functions no-unused-expression */
import { describe, it } from "mocha";
import { ProjektiPaths } from "../../src/files/ProjektiPath";

const { expect } = require("chai");

describe("ProjektiPaths", () => {
  it("should create yllapito paths successfully", function () {
    expect([
      new ProjektiPaths("123").yllapitoPath,
      new ProjektiPaths("123").vuorovaikutus({ vuorovaikutusNumero: 456 }).yllapitoPath,
      new ProjektiPaths("123").vuorovaikutus({ vuorovaikutusNumero: 456 }).kutsu.yllapitoPath,
      new ProjektiPaths("123").vuorovaikutus({ vuorovaikutusNumero: 456 }).aineisto.yllapitoPath,
      new ProjektiPaths("123").nahtavillaoloVaihe({ id: 7 }).yllapitoPath,
      new ProjektiPaths("123").hyvaksymisPaatosVaihe({ id: 8 }).yllapitoPath,
      new ProjektiPaths("123").hyvaksymisPaatosVaihe({ id: 8 }).paatos.yllapitoPath,
      new ProjektiPaths("123").jatkoPaatos1Vaihe({ id: 8 }).yllapitoPath,
      new ProjektiPaths("123").jatkoPaatos1Vaihe({ id: 8 }).paatos.yllapitoPath,
      new ProjektiPaths("123").jatkoPaatos2Vaihe({ id: 8 }).yllapitoPath,
      new ProjektiPaths("123").jatkoPaatos2Vaihe({ id: 8 }).paatos.yllapitoPath,
    ]).toMatchSnapshot();
  });

  it("should create yllapito full paths successfully", function () {
    expect([
      new ProjektiPaths("123").yllapitoFullPath,
      new ProjektiPaths("123").vuorovaikutus({ vuorovaikutusNumero: 456 }).yllapitoFullPath,
      new ProjektiPaths("123").vuorovaikutus({ vuorovaikutusNumero: 456 }).kutsu.yllapitoFullPath,
      new ProjektiPaths("123").vuorovaikutus({ vuorovaikutusNumero: 456 }).aineisto.yllapitoFullPath,
      new ProjektiPaths("123").nahtavillaoloVaihe({ id: 7 }).yllapitoFullPath,
      new ProjektiPaths("123").hyvaksymisPaatosVaihe({ id: 8 }).yllapitoFullPath,
      new ProjektiPaths("123").hyvaksymisPaatosVaihe({ id: 8 }).paatos.yllapitoFullPath,
      new ProjektiPaths("123").jatkoPaatos1Vaihe({ id: 8 }).yllapitoFullPath,
      new ProjektiPaths("123").jatkoPaatos1Vaihe({ id: 8 }).paatos.yllapitoFullPath,
      new ProjektiPaths("123").jatkoPaatos2Vaihe({ id: 8 }).yllapitoFullPath,
      new ProjektiPaths("123").jatkoPaatos2Vaihe({ id: 8 }).paatos.yllapitoFullPath,
    ]).toMatchSnapshot();
  });

  it("should create public projekti relative paths successfully", function () {
    expect([
      new ProjektiPaths("123").publicPath,
      new ProjektiPaths("123").vuorovaikutus({ vuorovaikutusNumero: 456 }).publicPath,
      new ProjektiPaths("123").vuorovaikutus({ vuorovaikutusNumero: 456 }).aineisto.publicPath,
      new ProjektiPaths("123").vuorovaikutus({ vuorovaikutusNumero: 456 }).kutsu.publicPath,
      new ProjektiPaths("123").nahtavillaoloVaihe({ id: 7 }).publicPath,
      new ProjektiPaths("123").hyvaksymisPaatosVaihe({ id: 8 }).publicPath,
      new ProjektiPaths("123").hyvaksymisPaatosVaihe({ id: 8 }).paatos.publicPath,
      new ProjektiPaths("123").jatkoPaatos1Vaihe({ id: 8 }).publicPath,
      new ProjektiPaths("123").jatkoPaatos1Vaihe({ id: 8 }).paatos.publicPath,
      new ProjektiPaths("123").jatkoPaatos2Vaihe({ id: 8 }).publicPath,
      new ProjektiPaths("123").jatkoPaatos2Vaihe({ id: 8 }).paatos.publicPath,
    ]).toMatchSnapshot();
  });

  it("should create public full paths successfully", function () {
    expect([
      new ProjektiPaths("123").publicFullPath,
      new ProjektiPaths("123").vuorovaikutus({ vuorovaikutusNumero: 456 }).publicFullPath,
      new ProjektiPaths("123").vuorovaikutus({ vuorovaikutusNumero: 456 }).aineisto.publicFullPath,
      new ProjektiPaths("123").vuorovaikutus({ vuorovaikutusNumero: 456 }).kutsu.publicFullPath,
      new ProjektiPaths("123").nahtavillaoloVaihe({ id: 7 }).publicFullPath,
      new ProjektiPaths("123").hyvaksymisPaatosVaihe({ id: 8 }).publicFullPath,
      new ProjektiPaths("123").hyvaksymisPaatosVaihe({ id: 8 }).paatos.publicFullPath,
      new ProjektiPaths("123").jatkoPaatos1Vaihe({ id: 8 }).publicFullPath,
      new ProjektiPaths("123").jatkoPaatos1Vaihe({ id: 8 }).paatos.publicFullPath,
      new ProjektiPaths("123").jatkoPaatos2Vaihe({ id: 8 }).publicFullPath,
      new ProjektiPaths("123").jatkoPaatos2Vaihe({ id: 8 }).paatos.publicFullPath,
    ]).toMatchSnapshot();
  });
});
