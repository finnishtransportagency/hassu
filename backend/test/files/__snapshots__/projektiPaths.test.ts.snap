// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProjektiPaths should create public full paths successfully 1`] = `
Array [
  "tiedostot/suunnitelma/123",
  "tiedostot/suunnitelma/123/suunnitteluvaihe/vuorovaikutus_456",
  "tiedostot/suunnitelma/123/suunnitteluvaihe/vuorovaikutus_456/aineisto",
  "tiedostot/suunnitelma/123/suunnitteluvaihe/vuorovaikutus_456/kutsu",
  "tiedostot/suunnitelma/123/nahtavillaolo",
  "tiedostot/suunnitelma/123/hyvaksymispaatos",
  "tiedostot/suunnitelma/123/hyvaksymispaatos/paatos",
  "tiedostot/suunnitelma/123/jatkopaatos1",
  "tiedostot/suunnitelma/123/jatkopaatos1/paatos",
  "tiedostot/suunnitelma/123/jatkopaatos2",
  "tiedostot/suunnitelma/123/jatkopaatos2/paatos",
]
`;

exports[`ProjektiPaths should create public projekti relative paths successfully 1`] = `
Array [
  "",
  "suunnitteluvaihe/vuorovaikutus_456",
  "suunnitteluvaihe/vuorovaikutus_456/aineisto",
  "suunnitteluvaihe/vuorovaikutus_456/kutsu",
  "nahtavillaolo",
  "hyvaksymispaatos",
  "hyvaksymispaatos/paatos",
  "jatkopaatos1",
  "jatkopaatos1/paatos",
  "jatkopaatos2",
  "jatkopaatos2/paatos",
]
`;

exports[`ProjektiPaths should create yllapito full paths successfully 1`] = `
Array [
  "yllapito/tiedostot/projekti/123",
  "yllapito/tiedostot/projekti/123/suunnitteluvaihe/vuorovaikutus_456",
  "yllapito/tiedostot/projekti/123/suunnitteluvaihe/vuorovaikutus_456/kutsu",
  "yllapito/tiedostot/projekti/123/suunnitteluvaihe/vuorovaikutus_456/aineisto",
  "yllapito/tiedostot/projekti/123/nahtavillaolo/7",
  "yllapito/tiedostot/projekti/123/hyvaksymispaatos/8",
  "yllapito/tiedostot/projekti/123/hyvaksymispaatos/8/paatos",
  "yllapito/tiedostot/projekti/123/jatkopaatos1/8",
  "yllapito/tiedostot/projekti/123/jatkopaatos1/8/paatos",
  "yllapito/tiedostot/projekti/123/jatkopaatos2/8",
  "yllapito/tiedostot/projekti/123/jatkopaatos2/8/paatos",
]
`;

exports[`ProjektiPaths should create yllapito paths successfully 1`] = `
Array [
  "",
  "suunnitteluvaihe/vuorovaikutus_456",
  "suunnitteluvaihe/vuorovaikutus_456/kutsu",
  "suunnitteluvaihe/vuorovaikutus_456/aineisto",
  "nahtavillaolo/7",
  "hyvaksymispaatos/8",
  "hyvaksymispaatos/8/paatos",
  "jatkopaatos1/8",
  "jatkopaatos1/8/paatos",
  "jatkopaatos2/8",
  "jatkopaatos2/8/paatos",
]
`;
